kernel/src/main/java/org/kframework/parser/inner/kernel/EarleyParser.java

```diff
@@ -0,0 +1,1023 @@
+   */
+  public EarleyParser(Module m, Scanner scanner, Sort startSymbol) {
+    this.scanner = scanner;
+
+    // compute metadata about grammar
+    sorts = getSorts(m);
+    productions = getProductions(m, scanner, startSymbol);
+    predictor = getPredictor();
+
+    //compute nullability
+    Set<LRItem> nullable = new HashSet<>();
+    List<Set<LRItem>> callers = getCallers();
+    this.nullable = new BitSet(sorts.size());
+    for (EarleyProduction prod : productions) {
+      markNullable(new LRItem(prod, 0), nullable, callers);
+    }
+
+    //compute first set
+    first = computeFirstSet();
+  }
+
+  private Map<Sort, Integer> getSorts(Module m) {
+    Map<Sort, Integer> sorts = new HashMap<>(m.allSorts().size());
+    Sort entrySort = Sort("<ENTRY>");
+    sorts.put(entrySort, 0);
+    sortList.add(entrySort);
+    Sort nullSort = Sort("<null>");
+    sorts.put(nullSort, 1);
+    sortList.add(nullSort);
+    int i = 2;
+    // some weirdness that turns out to be necessary due to how we implement MInt in RuleGrammarGenerator
+    for (SortHead sortHead : iterable(m.definedInstantiations().keySet())) {
+      Sort nonsenseSort = Sort(sortHead.name(), Seq(Sorts.K()));
+      if (!m.allSorts().contains(nonsenseSort)) {
+        sorts.put(nonsenseSort, i++);
+        sortList.add(nonsenseSort);
+      }
+    }
+    for (Sort s : iterable(m.allSorts())) {
+      sorts.put(s, i++);
+      sortList.add(s);
+    }
+    return sorts;
+  }
+
+  private List<EarleyProduction> getProductions(Module m, Scanner scanner, Sort startSymbol) {
+    final List<EarleyProduction> productions;
+    productions = new ArrayList<>(m.productions().size());
+    productions.add(new EarleyProduction(0, null, 0, Arrays.asList(new EarleyNonTerminal(sorts.get(startSymbol), sortList), new EarleyTerminal(scanner, 0)), sortList));
+    int index = 1;
+    for (Production prod : iterable(m.productions())) {
+      if (prod.params().size() != 0) {
+        continue;
+      }
+      List<EarleyProductionItem> items = new ArrayList<>(prod.items().size());
+      for (ProductionItem item : iterable(prod.items())) {
+        if (item instanceof Terminal) {
+          Terminal t = (Terminal) item;
+          if (t.value().isEmpty()) {
+            continue;
+          }
+        }
+        items.add(toEarley(item, scanner));
+      }
+      Optional<Production> original = prod.att().getOptional(Att.ORIGINAL_PRD(), Production.class);
+      Production prodToUse = prod;
+      if (original.isPresent()) {
+        prodToUse = original.get();
+      }
+      productions.add(new EarleyProduction(index++, prodToUse, sorts.get(prod.sort()), items, sortList));
+
+    }
+    return productions;
+  }
+
+  private EarleyProductionItem toEarley(ProductionItem item, Scanner scanner) {
+    if (item instanceof NonTerminal) {
+      Integer sort = sorts.get(((NonTerminal)item).sort());
+      if (sort == null) {
+        sort = 1; // <NULL>, a sort with no productions
```

Comment:
Shouldn't this be an error?
When can you find a sort in a module that doesn't exist already?

Reply:
this happens due to private productions, if you import a module privately, it is possible to have modules that rely on the existence of certain sorts despite those sort declarations not existing in that module. If you look, the old parser does something very similar in Grammar.java.

Reply:
Please add a comment with this explanation. This is not obvious from the code.